### ISSUE TITLE
Add new dark link option and improve inverted link accessibility documentation

### ISSUE THREAD

**Issue by bartaz** [canonical/vanilla-framework#4633]
As discussed in comments under https://github.com/canonical/vanilla-framework/issues/3350#issuecomment-1072448352 we need to improve how we handle links on dark backgrounds:

- introduce new class name for links on dark background (with new lighter blue colour)
- update current "inverted" link to show an underline (possibly rename it)
- update documentation to describe when to use new dark link and when to use "inverted" link

**Comment by lyubomir-popov** [comment]
I just filed a separate [issue](https://github.com/canonical/vanilla-framework/issues/4632), maybe close one of them and consolidate the info in the remaining one?

**Comment by bartaz** [comment]
Thanks @lyubomir-popov, I closed the other one and moved your comment here.

**Comment by bartaz** [comment]
In the dark theme make links underlined.

Inverted link (white underlined) is supposed to be only used on dark backgrounds we can't control (images)?

Blue underlined link is for solid dark backgrounds (new class for that).

**Comment by juanruitina** [comment]
We have received feedback [from a colleague](https://discourse.canonical.com/t/introducing-the-new-canonical-com/2204/3) that contrast is too low on dark theme links on canonical.com; I had the same intuition. Given contrast is heavily perceptual, users' perceptions should be taken into consideration.

The WCAG algorithm is known to throw false positives in many instances and for blues in particular ([one source](https://blogs.kent.ac.uk/webdev/2020/08/18/colours-contrast-and-accessibility/)), because blue has low luminance. As @lyubomir-popov mentioned already, the combination fails the APCA algorithm ([blue on graphite](https://contrast.tools?text=6699cc&background=262626), [purple on graphite](https://contrast.tools?text=a679d2&background=262626)), particularly given the very thin font weight we are using.

For context, here's [Github's own discussion around the topic](https://github.com/orgs/community/discussions/51233). They agree it's a tough one. I agree that "Visual task of differentiation is not all the same as the visual task of fluent readability" (a comment by the way made by the proponent of the APCA algorithm). It is a tradeoff, and I am on the side of ensuring readability, as differentiation can be achieved in other ways (e.g. positioning, underlining).

I acknowledge this is difficult to address, as we might need to choose between compliance and actual accessibility; this is something we have discussed a few times already. But wanted to document this nonetheless.

Some more reads: 
- https://github.com/w3c/low-vision-a11y-tf/issues/84
- [NASA research on designing with blue](https://colorusage.arc.nasa.gov/blue_2.php)

![canonical com_(Desktop)](https://github.com/vanilla/vanilla/assets/1409729/42a31cee-e7d5-4ebe-b261-1a0b80df474b)
![canonical com_(Desktop) (1)](https://github.com/vanilla/vanilla/assets/1409729/79ec1727-976b-4dd1-ae65-fb3f1bac0b95)

**Comment by juanruitina** [comment]
Following on the conversation with @lyubomir-popov, [I'm doing some more explorations on Figma](https://www.figma.com/file/PW5CNBzevLkewg71ENRTka/Dark-theme-link-experiments?type=design&node-id=0-1&mode=design&t=nEvNHx68eKf7M251-0), also comparing with how VS Code and GitHub handle this. VS Code combination clearly [fails](https://contrast-triangle.com/?bgColor=%23202020&linkColor=4DAAFC&textColor=989898) the [3:1 contrast ratio with surrounding text](https://www.w3.org/WAI/WCAG21/Techniques/general/G183), still seems to strike a good balance between readability and making links distinct. (Fun fact: I checked this by using the [web version of VS Code](https://vscode.dev/), so that's by all means subject to WCAG compliance :D)

I acknowledge that 3:1 contrast requirement with surrounding text is important, and I understand we don't want to use underlining for every link. I get it, can get a page such as the [home](https://ubuntu-com-13755.demos.haus/) visually heavy. But there are alternatives.

In many cases, links are already followed by a chevron to convey they are, indeed, links. Another low-risk mitigation we could explore is leveraging the prefers-contrast media query, to be extra careful with users that have set high contrast on their device's accessibility settings. For instance, we could add underlines to all links only in those cases:

```css
@media (prefers-contrast: more) {
  .is-graphite .inner-wrapper a:not([class^=p-button]), .is-dark .inner-wrapper a:not([class^=p-button]) {
      text-decoration: underline;
      text-decoration-thickness: 1px;
      /* perhaps with some offset so it looks better: */
      text-underline-offset: 0.2em;
  }

  .is-graphite .inner-wrapper a:not([class^=p-button]):hover, .is-dark .inner-wrapper a:not([class^=p-button]):hover {
      text-decoration-thickness: 2px;
  }
}
```

With chevrons or underlines, the 3:1 contrast is no longer needed, because function is already conveyed visually.

I'd personally feel inclined to use a lighter, crisper tone of blue such as `#79BCF9`(mine) or `#4DAAFC` (VS Code's), and use the mitigations above when possible. True, we might not be fully compliant with the letter of the standard, but we would be addressing real accessibility concerns flagged up by real users, and in line with the feedback [others](https://github.com/orgs/community/discussions/51233) are getting.

**Comment by lyubomir-popov** [comment]
@juanruitina can you clarify APCA's stance on contrast with surrounding text? we need to define what qualifies as "accessible" against both adjacent text and background. If we decide we want to switch to APCA compliance, that's fine, but let's understand what that means.

I guess we need to agree - if we achieve APCA "spot text" level, does that qualify in APCA's interpretation

![image](https://github.com/canonical/vanilla-framework/assets/2741678/b5f3391a-b2e2-47f3-9eb2-7e7a06243ad1)


some more benchmarking from me, with generic colours from vscode and elsewhere where dark themes are mature (e.g. chrome settings)
https://www.figma.com/file/H6MSsN3taoXXOJCPUbIImQ/Design-System-Library?type=design&node-id=7476%3A14558&mode=design&t=vmdB3pGXQE5wbpV1-1

![image](https://github.com/canonical/vanilla-framework/assets/2741678/13706240-3b5a-4bf2-8528-068b5406606a)


(note vscode's background is darker, giving them more clearance)






because there are shades that provide spot color against both our background 262626 and adjacent text white, for example:

**Comment by juanruitina** [comment]
I really think we should keep WCAG 2 as our reference. But we should be ready to make some (well documented) exceptions when we have strong evidence (feedback, benchmarking, research, in this case APCA) of the standard being insufficient.

APCA is not necessarily prescriptive around contrast with adjacent text. The APCA author has [opinions](https://github.com/orgs/community/discussions/51233#discussioncomment-5472621); I agree with him, but I won't claim that's a standard. In any case, I think APCA as a tool proves a contrast triangle with a blue link colour, a dark background and white adjacent text that is both WCAG 2 compliant and genuinely accessible is a chimera because of the low perceptual luminance of blue. The benchmarking we are doing now proves major companies are finding those issues and making compromises: "compliant" != accessible.

It's interesting to see Google Search also going for lighter, readable shades in the dark theme (`#99c3ff`, `#c58af9`), even with adjacent text, [failing the 3:1 requirement](https://contrast-triangle.com/?bgColor=%23202020&linkColor=%2399c3ff&textColor=%23e8e8e8):

![Screenshot 2024-04-24 at 20 11 31](https://github.com/canonical/vanilla-framework/assets/1409729/dd8840ce-f063-4c86-b51d-e3b0f85a850b)

In line with what I see VS Code and Google doing, I'm suggesting is to prioritise link colours that pass _both_ WCAG 2 and APCA on background, and have acceptable contrast with adjacent text that doesn't compromise the former, even if that strictly means failing the 3:1 requirement.

Ah, fascinating challenge.

**Comment by juanruitina** [comment]
I'm putting together a [spec](https://docs.google.com/document/d/1cC1397Mx4A6jIKAU75o3NtbkvECNuhQ7pnP9J5bGyMY/edit) collecting all the evidence.

**Comment by bartaz** [comment]
@juanruitina Do you know what's the status of this?

**Comment by juanruitina** [comment]
I have been working on a [spec doc](https://docs.google.com/document/d/1cC1397Mx4A6jIKAU75o3NtbkvECNuhQ7pnP9J5bGyMY/edit?tab=t.0) that covers this issue, just submitted it for review.

**Comment by bartaz** [comment]
Triaging: Seems relevant to both new and old architecture.
If colours are changed these new values can ideally be passed via tokens eventually to both new and current codebase. But even if not, old code base should be kept up to date with such accessibility issues.
